### PR TITLE
Convert parameter metadata into a front-end form

### PR DESCRIPTION
**Pull request by david-mears-2**
In this PR we go as far as using the metadata response from the R API (which contains information about the model parameters) to render a form. This front-end component (ParameterForm.vue) can currently cope with two types of parameterType: select, and globeSelect, both of which result in the same kind of form input. The resulting form input is rendered depending on the number and length of the options either as a select input or a set of radio buttons (stylised to look like a row of buttons).

This form can't yet be submitted - that's for the next PR.

We could have put metadata info into a pinia AppStore or some such, but since it's only used in one page (two components) for now, I haven't bothered with it yet; but I'm bearing in mind that this might be the simpler way forward. I will probably do that in the next PR too. This AppStore could also hold the 'versions' API response.

### CODE REVIEW

File: server/utils/defineRApiEventHandler.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
-  handler: EventHandler<T, D>,
-): EventHandler<T, D> =>
-  defineEventHandler<T>(async (event) => {
-    const response = await handler(event) as ApiResponse;
+// A wrapper for Nuxt's defineEventHandler that handles errors from the R API.
+export const defineRApiEventHandler = (
+  callback: (event: H3Event) => Promise<ApiResponse>,
+): EventHandler =>
+  defineEventHandler(async (event) => {
+    const response = await callback(event) as ApiResponse;
```
Comment (david-mears-2) [inline server/utils/defineRApiEventHandler.ts:9]: These changes were motivated by this comment: https://github.com/jameel-institute/daedalus-web-app/pull/10#discussion_r1713453435

File: components/ParameterForm.vue [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    // set the 'selected' attribute below.
+    // As a result, this list of options is recalculated each time an option is selected.
+    return { label: option.label, value: option.id, selected: option.id === formData.value![parameter.id] };
+  });
+};
+
+const optionsAreTerse = (parameter: Parameter) => {
+  const eachOptionIsASingleWord = parameter.options.filter((option) => {
+    return option.label.includes(" ");
+  }).length === 0;
```
Comment (EmmaLRussell) [inline components/ParameterForm.vue:181]:
```suggestion
  const eachOptionIsASingleWord = !parameter.options.some((option) => {
    return option.label.includes(" ");
  });
```

File: components/ParameterForm.vue [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+const largeScreen = ref(true);
+const breakpoint = 992; // CoreUI's "lg" breakpoint
+const setFieldSizes = () => {
+  if (window.innerWidth < breakpoint) {
+    largeScreen.value = false;
+  } else {
+    largeScreen.value = true;
+  }
+};
+
```
Comment (EmmaLRussell) [inline components/ParameterForm.vue:199]: This is repeated in SideBar - should really only be implemented in one place and ideally called in one place and then made available to all components. Could be put in the store and made reactive (in another ticket!)

File: docker/Dockerfile [hunk 4]
```diff
@@ -7,6 +7,7 @@ ENV NODE_ENV=production
 WORKDIR /src
 
 COPY . .
+RUN npm install -g npm@latest
```
Comment (EmmaLRussell) [inline docker/Dockerfile:10]: Did out of date npm cause problems in this ticket? Could update the base image if you want to use a more recent npm. Oh, is this related to rollup issue you describe below? Either way could maybe put a comment here because I'm a still a bit confused!

File: package.json [hunk 5]
```diff
@@ -70,7 +71,13 @@
...
     "1": "==============================",
     "2": "Vue version is overridden because of the issue described in the 'tip' in",
-    "3": "the installation section of https://pinia.vuejs.org/ssr/nuxt.html"
+    "3": "the installation section of https://pinia.vuejs.org/ssr/nuxt.html",
+    "4": "==============================",
+    "5": "@rollup/rollup-linux-x64-gnu is an optional dependency as a fix for the issue that Rollup",
+    "6": "describes here: https://github.com/rollup/rollup/blob/f83b3151e93253a45f5b8ccb9ccb2e04214bc490/native.js#L59",
+    "7": "and which occurred for us when doing an installation with npm on Docker on CI. Their suggested fix",
+    "8": "does not work for our use case, because removing package-lock.json prevents the use of `npm ci`, so instead",
+    "9": "we use the solution suggested here: https://github.com/vitejs/vite/discussions/15532#discussioncomment-10192839"
```
Comment (EmmaLRussell) [inline package.json:80]: Tbh, I find this trick for putting comments into package.json pretty unreadable. Given that the dependencies etc being referred to aren't actually in the vicinity of the relevant section of package.json anyway, I'm not sure there's much benefit in keeping them here. How about moving them to a section in the README file dedicated to npm dependency notes? That way you could make all these urls into working links!

File: types/daedalusApiResponseTypes.ts [hunk 6]
```diff
@@ -11,10 +11,32 @@ export interface ApiResponse<T extends object = object> {
...
+
+export interface Parameter {
+  id: string
+  label: string
+  parameterType: string
+  defaultOption: string | null
+  ordered: boolean
+  options: Array<ParameterOption>
+}
+export interface MetaData {
```
Comment (EmmaLRussell) [inline types/daedalusApiResponseTypes.ts:37]: I don't think MetaData needs a hump in the middle... :camel:

**Review by EmmaLRussell — COMMENTED** [review]
Looks great! Couple of minor comments. That coreui eccentricity is a bit concerning...

File: components/ParameterForm.vue [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+const icon = (parameter: Parameter) => {
+  switch (parameter.id) {
+    case "country":
+      return "cilGlobeAlt";
+    case "response":
+      return "cilShieldAlt";
+    case "vaccine":
+      return "cilIndustry";
+    case "pathogen":
+      return "cilBug";
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue:163]: > arguably the preferred icon should be provided as part of the metadata too
I'm not sure I see the benefit. If there is an abstraction layer, we'd still end up with the web app having a dict look up.

[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue:163]: You make a good point about the bug, I wasn't sure myself. I will look for a little virus icon that makes our style.

File: components/ParameterForm.vue [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+          class="parameter-icon"
+        />
+        <CFormSelect
+          :id="globeParameter.id"
+          v-model="formData[globeParameter.id]"
+          :data-test="`${globeParameter.id}-select`"
+          :label="globeParameter.label"
+          :aria-label="globeParameter.label"
+          :options="selectOptions(globeParameter)"
+          :size="largeScreen ? 'lg' : ''"
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue:80]: Good point

File: components/ParameterForm.vue [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    case "pathogen":
+      return "cilBug";
+    default:
+      return undefined;
+  }
+};
+
+const selectOptions = (parameter: Parameter) => {
+  return parameter.options.map((option: ParameterOption) => {
+    // Because the select component does not seem to honour the initial v-model value, we had to manually
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue:171]: Yes. I'm now replacing this component with native HTML elements!

File: components/ParameterForm.vue (hunk 3 above)
Comment (david-mears-2) [reply components/ParameterForm.vue:199]: I'll add this to the app-store plan!

File: docker/Dockerfile (hunk 4 above)
Comment (david-mears-2) [reply docker/Dockerfile:10]: I don't know if it was related, but it was something I did while trying to debug the rollup issue. While working on that, I updated my local npm version, which then put my machine 1 minor version ahead of the one used in the Docker container. It seemed like a good idea to use the latest npm available in both places?

File: package.json (hunk 5 above)
Comment (david-mears-2) [reply package.json:80]: That sounds better - they're pretty fiddly to write as well because of the line numbers!

File: components/ParameterForm.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue]: The colour's a little off, making my eye twitch, but I hope people can live with it for now because I couldn't quickly fix it (when you have inline svg you can change a fill attribute, but that doesn't seem to work when it's an svg loaded in an img tag)

File: components/ParameterForm.vue [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    } else {
+      // Don't set an empty value or there will be a disjoint between the select component and the formData object,
+      // since the select component will visually appear to have an option selected (the first), but the formData object will have
+      // an empty string.
+      const defaultOption = parameter.defaultOption || parameter.options[0].id;
+      accumulator[parameter.id] = ref(defaultOption);
+    }
+
+    return accumulator;
+  }, {} as { [key: string]: any }),
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue:115]: Turns out one of the 3rd-party components we're using, the radio button one, dictates that the type signature for the 'v-model' prop -- (which it translates into 'modelValue' behind the scenes!) -- is `string | boolean | unknown[] | undefined`.

From this, I will omit the Array type because we expect not to have multi-selects here.

I think I will keep the number type around as suggested, so it can be used by numeric inputs and we can do calculations with the value; and I'll just cast all values into strings when they are passed into the radio button v-model prop.

File: components/ParameterForm.vue [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+const props = defineProps<{
+  globeParameter: Parameter | undefined
+  metaData: MetaData | undefined
+  metadataFetchStatus: AsyncDataRequestStatus
+  metadataFetchError: FetchError | null
+}>();
+
+const formData = ref(
+  // Create a new object with keys set to the id values of the metaData.parameters array of objects, and all values set to refs with default values.
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue]: This seems to work:

```ts
ref(
    props.metadata.parameters.reduce((accumulator, parameter) => {
      accumulator[parameter.id] = parameter.defaultOption || parameter.options[0].id;
      return accumulator;
    }, {} as { [key: string]: string | number }),
  )
```

And swapping out `ref`/`reactive` still passes all tests. So we can avoid having individual `ref`s as I had before while still wrapping the whole thing in a `ref()` instead of a `reactive()`. Personally, I prefer the explicitness of having to type `.value`, which reminds me when I'm using something that's reactive.

File: components/ParameterForm.vue [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+      accumulator[parameter.id] = ref("Thailand");
+    } else {
+      // Don't set an empty value or there will be a disjoint between the select component and the formData object,
+      // since the select component will visually appear to have an option selected (the first), but the formData object will have
+      // an empty string.
+      const defaultOption = parameter.defaultOption || parameter.options[0].id;
+      accumulator[parameter.id] = ref(defaultOption);
+    }
+
+    return accumulator;
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue:114]: Fair. Can I entrust with you the task of setting 'Thailand' to be the default country for now, and reverting it to 'United Kingdom' after November? Note that at the moment metadata option values and ids are exactly the same for countries until we get in some ISO abbreviations - and defaultOption wants an id, so we'll have to update defaultOption from 'Thailand' to 'THA' at some point.

File: components/ParameterForm.vue [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+  }, {} as { [key: string]: any }),
+);
+
+const parametersOfTypeSelect = computed(() => {
+  if (props.metaData) {
+    return props.metaData.parameters.filter(parameter => parameter.parameterType === "select");
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/ParameterForm.vue]: Nice

**Review by EmmaLRussell — APPROVED** [review]
looks good! Are we supposed to include an attribution for that wikimedia svg?

**Comment by david-mears-2** [comment]
> Are we supposed to include an attribution for that wikimedia svg?

Well caught. Yes we are, "in any reasonable manner". I shall add a note to the ticket for the acknowledgements page (that sounds like less work than adapting a new svg with a more licentious license).

https://commons.wikimedia.org/wiki/File:Coronavirus_icon.svg

**Comment by EmmaLRussell** [comment]
> > Are we supposed to include an attribution for that wikimedia svg?
> 
> Well caught. Yes we are, "in any reasonable manner". I shall add a note to the ticket for the acknowledgements page (that sounds like less work than adapting a new svg with a more licentious license).
> 
> https://commons.wikimedia.org/wiki/File:Coronavirus_icon.svg

Thanks, sounds good
